fireplace/cards/utils.py

```diff
@@ -38,6 +40,46 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
+	"""
+	take a list of weights and a list of card pools and produce a random weighted
+	sample without replacement.
+	len(weights) == len(card_sets) (one weight per card set)
+	"""
+
+	# special case for single weight (generic random pick)
+	if len(card_sets) == 1:
+		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
+
+	# special case for all weights equal
+	if len(set(weights)) == 1:
```

Comment:
This case smells like too much of premature optimization to me.

We're only doing a small bit of extra book-keeping in the general case. Would you be willing to micro-bench this function with and without this shortcut to see how much of a difference it makes?